From 2a2bf431f3e85f9ebb3a8adfd71988dd469efcd3 Mon Sep 17 00:00:00 2001
From: Andreas Schmidt <andreas.schmidt@wago.com>
Date: Tue, 10 Jul 2018 17:06:44 +0200
Subject: [PATCH] swconfig: enable cut through for mv88e6321

PROFINET needs cut through functionality. This patch
enables cut through for ports 3 and 4 on mv88e6321
switch.

Signed-off-by: Andreas Schmidt <andreas.schmidt@wago.com>
Tested-by: Stefan Schwalowsky <stefan.schwalowsky@wago.com>
---
 drivers/net/dsa/mv88e6xxx/mv88e6xxx.h |  3 +++
 drivers/net/phy/mv88e6321-cfg.c       | 30 ++++++++++++++++++++++++++++++
 2 files changed, 33 insertions(+)

diff --git a/drivers/net/dsa/mv88e6xxx/mv88e6xxx.h b/drivers/net/dsa/mv88e6xxx/mv88e6xxx.h
index 4c7ec72..6077dcf 100644
--- a/drivers/net/dsa/mv88e6xxx/mv88e6xxx.h
+++ b/drivers/net/dsa/mv88e6xxx/mv88e6xxx.h
@@ -171,6 +171,9 @@
 #define PORT_OUT_FILTERED	0x13
 #define PORT_TAG_REGMAP_0123	0x18
 #define PORT_TAG_REGMAP_4567	0x19
+#define PORT_CUT_THROUGH	0x1f
+#define PORT_CUT_THROUGH_ENABLE			BIT(8)
+#define PORT_CUT_THROUGH_QUEUE_MASK		0xF
 
 #define GLOBAL_STATUS		0x00
 #define GLOBAL_STATUS_PPU_STATE BIT(15) /* 6351 and 6171 */
diff --git a/drivers/net/phy/mv88e6321-cfg.c b/drivers/net/phy/mv88e6321-cfg.c
index 40865d7..622f16a 100644
--- a/drivers/net/phy/mv88e6321-cfg.c
+++ b/drivers/net/phy/mv88e6321-cfg.c
@@ -359,12 +359,34 @@ static int mv88e6321_sw_force_speed_internal_phys(struct switch_dev *dev,
 	return ret;
 }
 
+static int mv88e6321_sw_enable_cut_through(struct switch_dev *dev,
+					   u8 sw_port_addr)
+{
+	int ret;
+	int reg_val;
+
+	ret = mdiobus_read(dev->mii_bus, sw_port_addr, PORT_CUT_THROUGH);
+	if (ret < 0)
+		goto out;
+
+	reg_val = (ret & ~PORT_CUT_THROUGH_QUEUE_MASK);
+	reg_val = reg_val | PORT_CUT_THROUGH_ENABLE | 0xF;
+	ret = mdiobus_write(dev->mii_bus, sw_port_addr, PORT_CUT_THROUGH,
+			    reg_val);
+
+out:
+	return ret;
+}
+
 static int mv88e6321_sw_setup_dflt_cfg(struct switch_dev *dev,
                                      const struct switch_attr *attr,
                                      struct switch_val *val)
 {
 	int ret;
 
+	if (!dev)
+		return -ENODEV;
+
 	ret = mv88e6321_sw_force_speed_external_phys(dev,
 						MV88E6321_EXTERNAL_PHY0);
 	if (ret < 0)
@@ -385,6 +407,14 @@ static int mv88e6321_sw_setup_dflt_cfg(struct switch_dev *dev,
 	if (ret < 0)
 		goto out;
 
+	ret = mv88e6321_sw_enable_cut_through(dev, MV88E6321_SW_PORT3_REG);
+	if (ret < 0)
+		goto out;
+
+	ret = mv88e6321_sw_enable_cut_through(dev, MV88E6321_SW_PORT4_REG);
+	if (ret < 0)
+		goto out;
+
 	ret = mv88e6321_setup_tcam(dev->mii_bus, of_root);
 
 	if (ret < 0)
-- 
2.7.4

